### ISSUE TITLE
Automatic testing makes less sense than expected (using SC 1.4.3 as an example)

### ISSUE THREAD

**Issue by JAWS-test** [w3c/wcag#998]
The document ["Challenges with Accessibility Guidelines Conformance and Testing"](https://w3c.github.io/wcag/conformance-challenges/) specifies for all WCAG SCs whether they can be tested automatically. I would argue that automatic testing can lead to reliable results in much fewer cases than is claimed there. A simple practical example:

All automatic test tools I know claim that they can reliably test SC 1.4.3 (text contrast) if the text is not on images.  

The [document](https://w3c.github.io/wcag/conformance-challenges/#contrast-minimum-success-criterion-1-4-3) seems to support this assertion:

> Automated tools can check the color of text against the background in most cases. However, when background images are used, automated tests aren't reliably able to check for minimum contrast of text against the image—especially if the image is a photograph or drawing where the text is placed over the image.

Unfortunately, that's not true. 

To understand the problem, [my example](https://codepen.io/jaws-test/pen/JjoRmrJ) can be used. 
The headings and their numbers (a-w) match those in the example

I tested it with 
* [WAVE](https://wave.webaim.org/extension/)
* [AXE](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd)
* [ANDI](https://www.ssa.gov/accessibility/andi/help/install.html)
* [Code Sniffer](https://squizlabs.github.io/HTML_CodeSniffer/)
* [ARC Toolkit](https://chrome.google.com/webstore/detail/arc-toolkit/chdkkkccnlfncngelccgbgfmjebmkmce)
* Chrome contrast measurement
* Accessiblity Tool from Firefox

### a) Text outside container with background-color

All tools except the Accessibility Tool of Firefox (F12) fail to determine the actual contrast of text and background. What they determine is only the contrast of the text and the container in which the text is located in the source code. But visually, the text can be outside its container.

The tools (except Firefox) show the text with sufficient contrasts as error and the text without sufficient contrasts as ok. 

Firefox also determines the contrast to a background graphic correctly, so that the restriction regarding background graphics is wrong. From a technical point of view it is possible to determine the contrast between text and background graphics. What cannot be determined is how the text shifts in relation to the background graphic if, for example, the display width or the font size changes. The problem does not only affect background graphics, but any text on background

### b) Text does not have enough contrast depending on screen width

What all tools can't determine automatically is how the text behaves to its container when the display size changes. The WCAG, however, [requires accessibility for each display size](https://www.w3.org/TR/WCAG21/#cc2).

In my opinion, the problem that certain display sizes cause WCAG violations that cannot be tested automatically (and that Firefox also fails to solve) affects many WCAG SCs. It's not just about media query breakpoints, as my example shows. 

However, these problems can hardly be found in the manual test, because nobody will test all possible combinations of display height and width with all SCs, although WCAG requires this. 

### c) Only one color defined

Another condition of SC 1.4.3, which is not checked by all tools, is that foreground and background color must be defined or both must not be defined (an error is when only one of them is defined). 

### d) Invisible text

The tools have difficulty recognizing invisible text and not reporting it as a problem. Most tools succeed in recognizing text with font size 0 as invisible. With the other 3 methods in the example, the tools fail.

### e) Disabled elements

Some tools will detect disabled items if they have been correctly marked as disabled with `disabled` or `aria-disabled` attribute. All tools fail to detect disabled items that have not been tagged correctly or if the `disabled` attribute cannot be used. For example, a link marked without href is a valid deactivated link. No tool can detect this.

### f) CSS content

No tool except Firefox measures contrasts in CSS content

### g) pure decoration

No tool can reliably recognize decorative content. Some tools do not measure contrasts when the content is marked with `aria-hidden`.
However, `aria-hidden` does not automatically mean decorative content.

### h) Font icon

No tool can distinguish between text and icons. Other contrast requirements apply to icons (SC 1.4.11).

### i) images of text

No tool recognizes text within graphics (although this would not be a problem with SVG graphics, for example)

### j) Different states

In addition, WCAG requires sufficient contrasts not only in the standard view, but also for interactive elements when they get the focus, when they are clicked, when they are activated, when they are mouse over, etc. 
All of these tools also fail to test these contrasts, although they are partly derivable from CSS.

### k) part of a picture

No tool can check the following exception "... are part of a picture that contains significant other visual content"

### l) logo 

No tool can check whether text is a logo

### m) aria-hidden

Some tools do not check the contrasts of content that is marked with `aria-hidden=true`

### n) outline

No tool recognizes contours

### o) conforming alternative version 

No tool can determine whether there is a switch to the conforming alternative version, whether this switch is accessible, whether the alternative version meets all WCAG criteria, and whether it is equivalent in content and function.

### p) subtitles

No tool can check the contrast of subtitles in videos

### q) Dynamic content

No tool can check the contrast of dynamically generated content on the page (e.g. tooltips, pop-ups, error messages, sub-menus) if it is generated when it is displayed. Few tools can only check this content if it is already (invisibly) present in the source code.

### r) Editorial content

No tool can check the contrast of future editorially created content, e.g. if the authors have the possibility to specify colors in the Rich Text Editor. 

### s) placeholder

No tool measures the contrast of placeholders 

### t) opacity

Some tools did not correctly measure the contrast of transparent areas

### u) Text in input field

Few tools do not check the contrast of text in input fields (e.g. Wave)

### v) colour gradient

No tool can calculate the contrast of CSS gradients. Some, but not all, tools indicate that the contrast must be checked manually.

### w) Form fields

Some tools measure the text contrast of form fields, even if they do not contain text (like `<input type=range>`).

### x) iFrames

Some tools do not measure text contrast in iFrames (e.g. Wave, Andi)

### y) Browser elements

Contrasts of texts generated by the browser (e.g. time display for `video` element) cannot be checked.

### z) Animation

All tools measure the contrast at the time of testing, making it difficult to determine a correct result for color animations.

**Comment by Myndex** [comment]
This list indicates a problem with testing tools not being comprehensive in their development, not that these things cannot be tested automatically.

Perhaps that points to a need for an SC that defines what must be completely tested automatically in order for a tool to claim automatic testing.

**Comment by JAWS-test** [comment]
The list contains problems that are only problems of the test tools and could be solved.

However, the list also contains problems that cannot be solved by the tools, but simply cannot be tested automatically.

In my opinion, however, the problems with the tools also show how difficult it is to test automatically.

If desired, I can make an assignment to the respective category. However, I have already described with the respective problem whether it is a problem of the tool or of the automatic test.

**Comment by Myndex** [comment]
> However, the list also contains problems that cannot be solved by the tools, but simply cannot be tested automatically.

Which ones? I didn't see anything that was specifically contrast related that also could not have automated testing. Some of your assertions are weird in that regard.

> In my opinion, the problems with the tools also show how difficult it is to test automatically.

The problems with the tools show how there is a lack of developers working to solve the issue.

> If desired, I can make an assignment to the respective category. However, I have already described with the respective problem whether it is a problem of the tool or of the automatic test.

???? Not clearly you haven't

**Comment by JAWS-test** [comment]
> Which ones? 

In my opinion the following things can never be tested automatically:

* No tool can reliably recognize decorative content
* No tool can distinguish between text and icons
* No tool recognizes text within graphics (OCR can try this, but will never work 100%)
* No tool can check the following exception "... are part of a picture that contains significant other visual content"
* No tool can check whether text is a logo
* Different states, if they are not defined in the CSS
* No tool can determine whether there is a switch to the conforming alternative version, whether this switch is accessible, whether the alternative version meets all WCAG criteria, and whether it is equivalent in content and function.
* No tool can check the contrast of subtitles in videos (if the subtitles are part of the video track)
* No tool can check the contrast of dynamically generated content on the page if it is generated when it is displayed
* No tool can check the contrast of future editorially created content
* No tool can recognize deactivated content if it is not marked with disabled, which is not possible e.g. with links.
* No tool can detect whether disabled content is relevant, so it still has to meet the contrast requirements.

I think: as extremely difficult to test is the most important thing: contrasts with all possible different display sizes.

And these are all just problems with 1.4.3. It becomes even more difficult with the contrast requirements of 1.4.11, which for the most part cannot be tested automatically.

**Comment by Myndex** [comment]
Most of your statements are untrue in terms of "cannot" as in "never will". Perhaps the current tools you use cannot do these things, but nearly everything you claim is "impossible" is already demonstrated as possible and indeed implemented / functioning for similar CV problems and solutions.

Just because you can't figure out how to do it does not mean it cannot be done. CV/ML is farther along than I think you may realize.

**Comment by alastc** [comment]
There are a couple of levels of automated checking, and I assume that @Myndex is correct that tools can be developed to test some/all of the aspects.

However, in terms of whether an SC/guideline could be fully automated across any website, I am sceptical. (I.e. There's a big difference between looking at page with a tool to assist evaluation, and an automated tool that can 'pass' a page without human intervention.)

It would probably help to keep the discussion to: What update would be useful to make to the contrast sections of the [Challenges doc](https://w3c.github.io/wcag/conformance-challenges/#contrast-minimum).

For the document the point is to outline what makes automated assessment difficult or impossible. In terms of categories of issue, I think the main ones are:
- The many ways of implementing text over a background, including page variations & image backgrounds.
- The aspects of the SC that derive from context rather than contrast. E.g. whether some text is incidental or part of a logo.

There are also things that the current tools struggle with (e.g. false positives for text that is never visible), but that isn't as relevant for this doc.

How about this as a starting point:
> Automated tools can check the color of text against the background in **many** cases. However, when background images are used, automated tests aren't reliably able to check for minimum contrast of text against the image—especially if the image is a photograph or drawing where the text is placed over the image. This **currently takes** human intervention to sample the text and its background to determine if the contrast meets the minimum requirement.
> **There are also exceptions in the SC where text is allowed to lack contrast, it takes a human to understand if something is a logo or purely decorative.**

**Comment by rg-viza** [comment]
some thoughts...

Jaws Inspect (also by Freedom Scientific) might help a great deal with automated jaws testing. I co-presented a symposium on it at the National Federation of the Blind with Karl Belanger.

What it does is it reads controls etc the way JAWS does, however instead of converting them to speech, it converts them to text (along with context) which can be exported as a spreadsheet. You can then feed this data to automated tools (such as phpUnit) as expected data and ensure your code is putting out good stuff by using assertions against the source compared to jaws inspect output.

As far as contrast goes you could (concievably) require RGB codes in the recommendation (instead of color names), then read and calculate the color distance based on the numbers.

Using ImageMagick or some other library, you could average the color of an image to an rgb average and calculate that color distance against text.

In a low-vision accessibility scenario, using background images is a bad idea to begin with. Any non-solid background is going to reduce contrast. The lower the contrast of the background image, the more this issue is alleviated, provided the RGB average of the image is of sufficient contrast to the text color. 

The image contrast can be calculated, then color averaged, then distanced from the text color. It should be trivial to come up with acceptable numbers here.

All images should have alt-text and the word "Logo" should, along with the text it represents, be in the alt text of a logo image.

I don't pretend to have all the answers (or any of them!), I just didn't see any of these considerations in the discussion yet, so figured I should add them.

Neil Davis

NFB Developer

**Comment by sajkaj** [comment]
Added multiple sentences beginning with: "However, there are several challenges with using current state■ to" [Appendix A.10: "Contrast Minimum](https://w3c.github.io/wcag/conformance-challenges/#contrast-minimum).
